File: scripts/agents/companion/trim_static_tails.py
#!/usr/bin/env python3
"""Trim static tails from loop clips.

Analyzes frame-to-frame scene change scores to find where motion ends,
then trims each clip to remove the frozen tail. Keeps a 0.3s buffer
so the cut doesn't feel abrupt.

Overwrites originals after backing up.
"""
import subprocess
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))
from config import IDLE_LOOPS_DIR

LOOPS_DIR = IDLE_LOOPS_DIR
MOTION_THRESHOLD = 0.004  # rolling avg scene score below this = static
WINDOW_FRAMES = 24        # 1 second rolling window at 24fps
TAIL_BUFFER = 0.3         # keep 0.3s after last motion for smooth end
MIN_TRIM = 0.5            # only trim if saving more than 0.5s


def get_duration(path):
    r = subprocess.run(
        ["ffprobe", "-v", "error", "-show_entries", "format=duration",
         "-of", "csv=p=0", str(path)],
        capture_output=True, text=True, timeout=15,
    )
    return float(r.stdout.strip())


def find_trim_point(path):
    """Return the time (seconds) to trim to, or None if no trim needed."""
    duration = get_duration(path)

    r = subprocess.run(
        ["ffmpeg", "-i", str(path),
         "-vf", "select='gte(scene,0)',metadata=print:file=/dev/stdout",
         "-an", "-f", "null", "-"],
        capture_output=True, text=True, timeout=60,
    )

    scores = []
    for line in r.stdout.splitlines():
        if "scene_score" in line:
            val = line.split("=")[-1]
            try:
                scores.append(float(val))
            except ValueError:
                pass

    if len(scores) < WINDOW_FRAMES * 2:
        return None

    fps = len(scores) / duration
    w = WINDOW_FRAMES

    # Rolling mean
    means = [sum(scores[i:i+w]) / w for i in range(len(scores) - w)]

    # Find last frame with significant motion
    last_active = 0
    for i, m in enumerate(means):
        if m > MOTION_THRESHOLD:
            last_active = i + w

    trim_time = min(last_active / fps + TAIL_BUFFER, duration)
    saved = duration - trim_time

    if saved < MIN_TRIM:
        return None

    return trim_time


def trim_clip(path, trim_time):
    """Trim clip to trim_time seconds, overwriting original."""
    tmp = path.with_suffix(".trimmed.mp4")
    subprocess.run(
        ["ffmpeg", "-y", "-i", str(path),
         "-t", f"{trim_time:.3f}",
         "-c:v", "libx264", "-crf", "18", "-pix_fmt", "yuv420p",
         "-an", str(tmp)],
        capture_output=True, timeout=60,
    )
    if tmp.exists() and tmp.stat().st_size > 0:
        # Backup original
        bak = path.with_suffix(".pretrim.mp4")
        if not bak.exists():
            path.rename(bak)
        else:
            path.unlink()
        tmp.rename(path)
        return True
    return False


def main():
    clips = sorted(LOOPS_DIR.glob("loop_*.mp4"))
    if not clips:
        print("No loop clips found")
        return

    print(f"Analyzing {len(clips)} clips...\n")

    trimmed = 0
    for clip in clips:
        duration = get_duration(clip)
        trim_time = find_trim_point(clip)

        if trim_time is None:
            print(f"  {clip.name}: {duration:.1f}s — no trim needed")
            continue

        saved = duration - trim_time
        print(f"  {clip.name}: {duration:.1f}s -> {trim_time:.1f}s (trimming {saved:.1f}s)")

        if trim_clip(clip, trim_time):
            new_dur = get_duration(clip)
            print(f"    done ({new_dur:.1f}s)")
            trimmed += 1
        else:
            print(f"    FAILED")

    print(f"\nTrimmed {trimmed}/{len(clips)} clips.")
    if trimmed:
        print("Originals backed up as .pretrim.mp4")


if __name__ == "__main__":
    main()

File: src/main/jobs/trim-static-tails.ts
/**
 * Trim static tails from generated video loop clips.
 * Port of scripts/agents/companion/trim_static_tails.py.
 *
 * Analyzes frame-to-frame scene change scores to find where motion ends,
 * then trims each clip to remove the frozen tail. Keeps a 0.3s buffer
 * so the cut doesn't feel abrupt.
 *
 * Overwrites originals after backing up to .pretrim.mp4.
 *
 * Requires: ffmpeg and ffprobe installed.
 */

import { spawnSync } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';
import { getConfig } from '../config';
import { registerJob } from './index';
import { createLogger } from '../logger';

const log = createLogger('trim-static-tails');

// ---------------------------------------------------------------------------
// Constants - matching the Python source exactly
// ---------------------------------------------------------------------------

const MOTION_THRESHOLD = 0.004; // rolling avg scene score below this = static
const WINDOW_FRAMES = 24; // 1 second rolling window at 24fps
const TAIL_BUFFER = 0.3; // keep 0.3s after last motion for smooth end
const MIN_TRIM = 0.5; // only trim if saving more than 0.5s

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function getVideoDuration(videoPath: string): number {
  const result = spawnSync(
    'ffprobe',
    [
      '-v', 'error',
      '-show_entries', 'format=duration',
      '-of', 'csv=p=0',
      videoPath,
    ],
    { encoding: 'utf-8', timeout: 15_000 },
  );
  return parseFloat(result.stdout.trim());
}

/**
 * Analyze frame-to-frame scene change scores and find the optimal trim point.
 * Returns the time in seconds to trim to, or null if no trim is needed.
 */
function findTrimPoint(videoPath: string): number | null {
  const duration = getVideoDuration(videoPath);

  // Extract scene change scores for every frame
  const result = spawnSync(
    'ffmpeg',
    [
      '-i', videoPath,
      '-vf', "select='gte(scene,0)',metadata=print:file=/dev/stdout",
      '-an',
      '-f', 'null',
      '-',
    ],
    { encoding: 'utf-8', timeout: 60_000 },
  );

  // Parse scene_score values from ffmpeg output
  const scores: number[] = [];
  for (const line of (result.stdout || '').split('\n')) {
    if (line.includes('scene_score')) {
      const val = line.split('=').pop();
      if (val) {
        const num = parseFloat(val);
        if (!isNaN(num)) {
          scores.push(num);
        }
      }
    }
  }

  // Need enough frames for meaningful analysis
  if (scores.length < WINDOW_FRAMES * 2) {
    return null;
  }

  const fps = scores.length / duration;

  // Compute rolling mean of scene scores
  const means: number[] = [];
  for (let i = 0; i <= scores.length - WINDOW_FRAMES; i++) {
    let sum = 0;
    for (let j = i; j < i + WINDOW_FRAMES; j++) {
      sum += scores[j];
    }
    means.push(sum / WINDOW_FRAMES);
  }

  // Find the last frame with significant motion
  let lastActive = 0;
  for (let i = 0; i < means.length; i++) {
    if (means[i] > MOTION_THRESHOLD) {
      lastActive = i + WINDOW_FRAMES;
    }
  }

  const trimTime = Math.min(lastActive / fps + TAIL_BUFFER, duration);
  const saved = duration - trimTime;

  if (saved < MIN_TRIM) {
    return null;
  }

  return trimTime;
}

/**
 * Trim a video clip to the given time, overwriting the original after backup.
 */
function trimClip(videoPath: string, trimTime: number): boolean {
  const tmpPath = videoPath.replace(/\.mp4$/, '.trimmed.mp4');

  spawnSync(
    'ffmpeg',
    [
      '-y',
      '-i', videoPath,
      '-t', trimTime.toFixed(3),
      '-c:v', 'libx264',
      '-crf', '18',
      '-pix_fmt', 'yuv420p',
      '-an',
      tmpPath,
    ],
    { stdio: 'pipe', timeout: 60_000 },
  );

  if (!fs.existsSync(tmpPath) || fs.statSync(tmpPath).size === 0) {
    // Clean up failed attempt
    try { fs.unlinkSync(tmpPath); } catch { /* ignore */ }
    return false;
  }

  // Backup original
  const bakPath = videoPath.replace(/\.mp4$/, '.pretrim.mp4');
  if (!fs.existsSync(bakPath)) {
    fs.renameSync(videoPath, bakPath);
  } else {
    fs.unlinkSync(videoPath);
  }

  // Replace with trimmed version
  fs.renameSync(tmpPath, videoPath);
  return true;
}

// ---------------------------------------------------------------------------
// Main entry point
// ---------------------------------------------------------------------------

export async function runTrimStaticTails(agentName: string): Promise<string> {
  const config = getConfig();
  config.reloadForAgent(agentName);

  const loopsDir = config.IDLE_LOOPS_DIR;

  if (!fs.existsSync(loopsDir)) {
    return 'Skipped: loops directory does not exist';
  }

  // Find all loop_*.mp4 files
  const clips = fs
    .readdirSync(loopsDir)
    .filter((f) => f.startsWith('loop_') && f.endsWith('.mp4'))
    .sort()
    .map((f) => path.join(loopsDir, f));

  if (clips.length === 0) {
    return 'No loop clips found';
  }

  log.info(`Analyzing ${clips.length} clips...`);

  let trimmed = 0;
  const results: string[] = [];

  for (const clip of clips) {
    const basename = path.basename(clip);
    const duration = getVideoDuration(clip);
    const trimTime = findTrimPoint(clip);

    if (trimTime === null) {
      log.info(`${basename}: ${duration.toFixed(1)}s - no trim needed`);
      results.push(`${basename}: no trim`);
      continue;
    }

    const saved = duration - trimTime;
    log.info(
      `${basename}: ${duration.toFixed(1)}s -> ${trimTime.toFixed(1)}s ` +
        `(trimming ${saved.toFixed(1)}s)`,
    );

    if (trimClip(clip, trimTime)) {
      const newDur = getVideoDuration(clip);
      log.info(`  done (${newDur.toFixed(1)}s)`);
      results.push(`${basename}: ${duration.toFixed(1)}s -> ${newDur.toFixed(1)}s`);
      trimmed++;
    } else {
      log.error(`  ${basename}: trim FAILED`);
      results.push(`${basename}: FAILED`);
    }
  }

  const summary = `Trimmed ${trimmed}/${clips.length} clips`;
  if (trimmed > 0) {
    log.info(`${summary}. Originals backed up as .pretrim.mp4`);
  } else {
    log.info(summary);
  }

  return `${summary}. ${results.join('; ')}`;
}

// ---------------------------------------------------------------------------
// Job registration
// ---------------------------------------------------------------------------

registerJob({
  name: 'trim-static-tails',
  description: 'Trim static/frozen tails from generated video loop clips using scene analysis',
  gates: [
    () => {
      // Check if ffprobe is available
      const result = spawnSync('which', ['ffprobe'], { encoding: 'utf-8', timeout: 5000 });
      if (result.status !== 0) return 'ffprobe not found';
      return null;
    },
  ],
  run: async () => {
    const config = getConfig();
    return runTrimStaticTails(config.AGENT_NAME);
  },
});
